unblob/handlers/filesystem/yaffs/utils.py

```diff
@@ -0,0 +1,574 @@
+    file_size=0,
+    start_offset=0,
+    chunks=[],
+)
+
+
+SPARE_START_BIG_ENDIAN_ECC = b"\x00\x00\x10\x00"
+SPARE_START_BIG_ENDIAN_NO_ECC = b"\xFF\xFF\x00\x00\x10\x00"
+SPARE_START_LITTLE_ENDIAN_ECC = b"\x00\x10\x00\x00"
+SPARE_START_LITTLE_ENDIAN_NO_ECC = b"\xFF\xFF\x00\x10\x00\x00"
+PAGE_START_LITTLE_ENDIAN = b"\x01\x00\x00\x00\x01\x00\x00\x00\xff\xff"
+PAGE_START_BIG_ENDIAN = b"\x00\x00\x00\x01\x00\x00\x00\x01\xff\xff"
+
+# 00 00 c0 ff ff ff 01 00  9a aa a7 a4 c1 59 aa ab
+# 00 00 c0 ff ff ff 01 01  03 ff 0f 00 c1 59 aa ab
+# 00 00 c0 ff ff ff 02 01  0f ff f3 0c c1 fc ff 03
+
+# 00 00 03 ff ff ff 00 00  95 aa a7 40 3f 56 aa ab
+# 00 00 03 ff ff ff 00 40  0c ff 0f 40 83 56 aa ab
+# 00 00 03 ff ff ff 00 40  00 ff f3 80 bf fc ff 03
+
+
+class YAFFSParser:
+    def __init__(self, file: File, config: Optional[YAFFSConfig] = None):
+        self.file_entries = Tree()
+        self.file_entries.create_node(ROOT, 1)
+        self.file = file
+        self._struct_parser = StructParser(C_DEFINITIONS)
+        self.file.seek(0, io.SEEK_END)
+        self.eof = self.file.tell()
+        self.file.seek(0, io.SEEK_SET)
+        self.end_offset = -1
+        if config is None:
+            self.config = self.auto_detect()
+            logger.debug("auto-detected config", config=self.config)
+        else:
+            self.config = config
+
+    def bruteforce(self) -> YAFFSConfig:
+        # let's do a trick here
+        entries = []
+        count = 0
+        config = YAFFSConfig(
+            endianness=Endian.LITTLE, page_size=-1, spare_size=-1, ecc=False
+        )
+        for page_size in VALID_PAGE_SIZES:
+            for spare_size in VALID_SPARE_SIZES:
+                entries = []
+                for i in range(0, self.eof // (page_size + spare_size)):
+                    start = (page_size + spare_size) * i
+                    entries.append(self.file[start : start + 10])
+
+                le_count = sum(
+                    [entry.startswith(PAGE_START_LITTLE_ENDIAN) for entry in entries]
+                )
+                be_count = sum(
+                    [entry.startswith(PAGE_START_BIG_ENDIAN) for entry in entries]
+                )
+
+                if le_count > count or be_count > count:
+                    config.endianness = (
+                        Endian.LITTLE if le_count > be_count else Endian.BIG
+                    )
+                    config.page_size = page_size
+                    config.spare_size = spare_size
+                    count = max([le_count, be_count])
+
+        if config.page_size == -1:
+            raise InvalidInputFormat("Can't find YAFFS config through bruteforce.")
+
+        return config
+
+    def auto_detect(self) -> YAFFSConfig:
+        """Auto-detect page_size, spare_size, and ECC using known signatures."""
+        config = None
+        page_size = 0
+        for page_size in VALID_PAGE_SIZES:
+            if self.file[page_size:].startswith(SPARE_START_LITTLE_ENDIAN_ECC):
```

Comment:
`self.file[page_size:]` could be very big and this result ina  copy (I think) we should specify and end offset as wll

Reply:
Indeed. Now only checking as many bytes as required.